### DISCUSSION TITLE
Is there any reason why using [Parameter()] on It block parameters in parameterised tests, breaks the tests?

### DISCUSSION THREAD

**Discussion by plastikfan** [pester/Pester#2043] [Q&A] [answered]
Taking the test suite as documented in the Pester readme:

```powershell
Describe 'Pester readme tests' -Tag 'Current' {
  BeforeAll {
    # your function
    function Get-Planet ([string]$Name = '*') {
      $planets = @(
        @{ Name = 'Mercury' }
        @{ Name = 'Venus' }
        @{ Name = 'Earth' }
        @{ Name = 'Mars' }
        @{ Name = 'Jupiter' }
        @{ Name = 'Saturn' }
        @{ Name = 'Uranus' }
        @{ Name = 'Neptune' }
      ) | foreach { [PSCustomObject]$_ }

      $planets | where { $_.Name -like $Name }
    }
  }

  # Pester tests
  Describe 'Get-Planet' {
    It "Given no parameters, it lists all 8 planets" {
      $allPlanets = Get-Planet
      $allPlanets.Count | Should -Be 8
    }

    Context "Filtering by Name" {
      It "Given valid -Name '<Filter>', it returns '<Expected>'" -TestCases @(
        @{ Filter = 'Earth'; Expected = 'Earth' }
        @{ Filter = 'ne*'  ; Expected = 'Neptune' }
        @{ Filter = 'ur*'  ; Expected = 'Uranus' }
        @{ Filter = 'm*'   ; Expected = 'Mercury', 'Mars' }
      ) {
        param (
          [Parameter()]
          $Filter,

          [Parameter()]
          $Expected
        )

        $planets = Get-Planet -Name $Filter
        $planets.Name | Should -Be $Expected
      }

      It "Given invalid parameter -Name 'Alpha Centauri', it returns `$null" {
        $planets = Get-Planet -Name 'Alpha Centauri'
        $planets | Should -Be $null
      }
    }
  }
}
```

In this example above, all I have done is to use the ___[Parameter()]___ attribute on the parameters passed into the It block, but for some strange reason, it breaks the tests with these parameter binding errors:

```
[-] Pester README tests.Get-Planet.Filtering by Name.Given valid -Name 'Earth', it returns 'Earth' 3ms (3ms|1ms)
 ParameterBindingException: A parameter cannot be found that matches parameter name '_'.

[-] Pester README tests.Get-Planet.Filtering by Name.Given valid -Name 'ne*', it returns 'Neptune' 2ms (1ms|1ms)
 ParameterBindingException: A parameter cannot be found that matches parameter name '_'.

[-] Pester README tests.Get-Planet.Filtering by Name.Given valid -Name 'ur*', it returns 'Uranus' 1ms (1ms|1ms)
 ParameterBindingException: A parameter cannot be found that matches parameter name '_'.

[-] Pester README tests.Get-Planet.Filtering by Name.Given valid -Name 'm*', it returns 'Mercury Mars' 2ms (1ms|1ms)
 ParameterBindingException: A parameter cannot be found that matches parameter name '_'.

```

I discovered this whilst writing some of my own parameterised tests that were in-explicably failing and decided to reproduce this error on other code, to rule out anything I may have been doing wrong accidentally.

**Comment by fflaten** [comment, marked as answer]
It's because the use of ParameterAttribute will make the scriptblock advanced (like `[CmdLetBinding()]`) which doesn't allow arguments by default. So it fails to bind some of the hard-coded values Pester passes to the `It`-scriptblock, in this case the object for the current testcase.

I'm not really sure why the param-block is in the README-example tbh. You don't really need it for normal usage. All the testcase parameters are made available both as named variables and the current testcase-object/dictionary itself can be accessed through `$_`.

Any reason you need the ParameterAttributes? If not I'd remove them or maybe even the whole param-block.

**Reply by plastikfan** [reply]
Hi @fflaten, thanks for your reply, which makes sense. The readme doesn't include the Parameter attribute, I was just testing it with the attribute because that is how I wrote my own parameterised test cases, which were failing due to the binding error. I was trying to figure this out for quite some time, stepping into the Pester code, to no avail. I didnt know that the error was due to the Parameter attribute. When I wrote my test suite, I just used the attribute just out of habit, not realising the error it would cause in this case. All I saw was the binding error and I didnt make the connection that it was because of the attribute.

PS: you do need the param block when running InModuleScope (not illustrated here, but in my test suite I am using InModuleScope), otherwise the testcase parameters (like \<Filter> and \<Expected> in this case) would not be bound inside the It block.

**Reply by nohwnd** [reply]
> PS: you do need the param block when running InModuleScope (not illustrated here, but in my test suite I am using InModuleScope), otherwise the testcase parameters (like <Filter> and <Expected> in this case) would not be bound inside the It block.

That should not be the case, please file a bug with an example.

**Reply by fflaten** [reply]
> PS: you do need the param block when running InModuleScope (not illustrated here, but in my test suite I am using InModuleScope), otherwise the testcase parameters (like \<Filter> and \<Expected> in this case) would not be bound inside the It block.

This is fixed in 5.3 (#1957). If you got time, please try the beta to see if it works as expected.

**Reply by plastikfan** [reply]
Argh ok. I've just realised I haven't updated my version of pester for a while, I'm still running 5.1.1, but just to give you a hint as to what InModuleScope code I'm running:

```Powershell
Describe 'convert-ActionResult' {
  BeforeAll {
    Get-Module ims | Remove-Module -Force;
    Import-Module .\ims.psm1 `
      -ErrorAction 'stop' -DisableNameChecking -Force;
  }

  Context 'given: <Scenario> <Result>' {
    It 'should: return result with Success = <ExpectedSuccess>' -TestCases @(
      # string results
      #
      @{
        Scenario        = 'Positive string result';
        Result          = 'affirmative';
        ExpectedSuccess = $true;
      }

      , @{
        Scenario        = 'Negative empty string result';
        Result          = [string]::Empty;
        ExpectedSuccess = $false;
      }

    ) {
      InModuleScope ims -Parameters @{ Scenario = $Scenario; Result = $Result; ExpectedSuccess = $ExpectedSuccess; } {
        param(
          [string]$Scenario,
          [object]$Result,
          [boolean]$ExpectedSuccess
        )

        [PSCustomObject]$actionResult = convert-ActionResult -Result $Result;
        $($null -ne ${actionResult}.Success) | Should -BeTrue -Because $Scenario;
        $actionResult.Success | Should -Be $ExpectedSuccess;

        $($null -ne ${actionResult}.Payload) -xor $(
          $null -ne ${actionResult}.FailedReason
        ) | Should -BeTrue -Because $Scenario;
      }
    }
  }
}
```

The command under test is __convert-ActionResult__, which is internal to the module I've called __ims__.

If I remove the parameter passing in the __InModuleScope__ statement, then the template parameters \<Scenario>, \<Result> and \<ExpectedSuccess> are all unbound. I used this technique as a result of reading some module/data-driven test documentation on pester.dev, quite some time ago and it served me well ever since.

I'm going to update my version of pester and retest.

**Reply by plastikfan** [reply]
I've just updated to version 5.3.0-beta, and tested and so far, there's no difference in behaviour. I'll continue this discussion on #1957.

**Reply by fflaten** [reply]
> If I remove the parameter passing in the __InModuleScope__ statement, then the template parameters \<Scenario>, \<Result> and \<ExpectedSuccess> are all unbound. 

5.3 removes the requirement for the param-block inside the scriptblock. You still need to specify `InModuleScope MyModule -Parameters ...`.

That's because we're passing data to another state (inside the module), so we need to be specific to not pollute the module.

**Reply by plastikfan** [reply]
✨ BINGO!!!! 😃
